### PR TITLE
cpu/saml1x, saml2x, samd5x: select buck voltage regulator when possible

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

On MCUs newer than `samd2x` we can select between LDO and buck converter for on-chip voltage conversion.

As the buck converter is much more efficient, select it by default.

On `same54-xpro` with `examples/default` this gives the following IDLE results:
    
**before:** 4.20 mA
 **after:** 2.06 mA

On `saml21-xpro` with `examples/default` this gives
    
**before:** 750 µA
 **after:** 385 µA

I could not test saml1x due to a lock of hardware.

### Testing procedure

No change in behavior is expected.

### Issues/PRs references
#13415 for further power saving

### CODE REVIEW

**Comment by jue89** [comment]
I noticed that the SAML21 datasheet [1] mentions the buck regulator in its erata:

> 51.1.5 – When VDDCORE is supplied by the BUCK converter in performance
> level 0, the chip cannot wake-up from standby mode because the
> VCORERDY status is stuck at 0.

They propose to switch back to LDO before entering STANDBY mode. This is another use case for the proposed callback concept of #13415.

[1] http://ww1.microchip.com/downloads/en/DeviceDoc/60001477A.pdf

**Comment by dylad** [comment]
Awesome ! This is something I had in mind for quite some time.
I'll run this PR on hardware when I'm back to office next week.

Regarding SAML21, I am a bit puzzled because I was hoping to merge ##10075 before I my retirement. According to errata, these two aren't compatible even with lastest die.

**Comment by benpicco** [comment]
> They propose to switch back to LDO before entering STANDBY mode. This is another use case for the proposed callback concept of #13415.

It is indeed! And samd5x has an errata wrt standby too where the DFLL needs to be re-initialized, so I can make use of this too. 

As for saml21, I could not observe the device getting stuck on wake from standby, so I can't test if the LDO switching actually works.

There is another errata where the main clock will not switch off in STANDBY if ` PM.STDBYCFG.VREGSMOD` is set to 2. It is suggested to switch it to OSCULP32k as a workaround.
This only affects revision B, I don't have that silicon.

File: cpu/saml21/cpu.c [hunk 1]
```diff
@@ -82,6 +82,28 @@ uint32_t sam0_gclk_freq(uint8_t id)
     }
 }
 
+void cpu_pm_cb_enter(int deep)
+{
+    if (deep) {
```
Comment (jue89) [inline cpu/saml21/cpu.c]: `deep` is always `0`. I think you should check `PM->SLEEPCFG.bit.SLEEPMODE`.

Comment (benpicco) [reply cpu/saml21/cpu.c]: Good catch - maybe we should just set deep for consistency.

**Comment by jue89** [comment]
> There is another errata where the main clock will not switch off in STANDBY if ` PM.STDBYCFG.VREGSMOD` is set to 2. It is suggested to switch it to OSCULP32k as a workaround.
> This only affects revision B, I don't have that silicon.

RIOT isn't modifying this register. Its reset value is 0x0. I would recommend to include an `assert()` in `cpu_pm_cb_enter()` checking if `PM.STDBYCFG.VREGSMOD != 0x2`. If someone is writing 0x2 to this register, `DEVELHELP` will spit out a warning.

**Comment by benpicco** [comment]
I mean the fix would be as simple as 

```patch
--- a/cpu/saml21/cpu.c
+++ b/cpu/saml21/cpu.c
@@ -92,6 +92,13 @@ void cpu_pm_cb_enter(int deep)
         /* select LDO regulator */
         SUPC->VREG.bit.SEL = 0;
         while (!SUPC->STATUS.bit.VREGRDY) {}
+
+        /* If the PM.STDBYCFG.VREGSMOD field is set to 2 (low-power configuration),
+           the oscillator source driving the GCLK_MAIN clock will still be running
+           in Standby mode causing extra consumption. */
+        if (PM->STDBYCFG.reg & 0x80) {
+            _gclk_setup(SAM0_GCLK_MAIN, GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_OSCULP32K);
+        }
     }
 }
 
@@ -101,6 +108,10 @@ void cpu_pm_cb_leave(int deep)
         /* select buck voltage regulator */
         SUPC->VREG.bit.SEL = 1;
         while (!SUPC->STATUS.bit.VREGRDY) {}
+
+        if (PM->STDBYCFG.reg & 0x80) {
+            _gclk_setup(SAM0_GCLK_MAIN, GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_OSC16M);
+        }
     }
 }
```

But I guess as long as we are not using it, there is no need to work around it.

**Comment by jue89** [comment]
> But I guess as long as we are not using it, there is no need to work around it.

I proposed using `assert()` because it doesn't cost any CPU cycles in production but would warn the user during development.

**Comment by jue89** [comment]
I tried your PR wit the samr30-xpro board. The commit 999a1a7 breaks samr30 CPUs.

Its datasheet shows that the `SUPC->VREG->SEL` bit is not available for the SAMR30. I would suggest to add a pre-processor `#ifndef CPU_SAMR30`. (Unfortunately, `SUPC->VREG.bit.SEL` is defined, nevertheless. So compile tests doesn't fail here ...)

**Comment by benpicco** [comment]
I was wondering if this is also true for samr34, but the data sheet just refers to saml21 for details, so I guess not.

File: cpu/samd5x/cpu.c [hunk 2]
```diff
@@ -152,6 +173,10 @@ void cpu_init(void)
     /* initialize the Cortex-M core */
     cortexm_init();
 
+    /* select buck voltage regulator */
+    SUPC->VREG.bit.SEL = 1;
+    while (!SUPC->STATUS.bit.VREGRDY) {}
```
Comment (jue89) [inline cpu/samd5x/cpu.c]: This must be excluded for the SAMR30. As already reported, the `SEL` bit is not defined for this CPU.

Comment (jue89) [reply cpu/samd5x/cpu.c]: With your fixup, `tests/periph_pm` is now working like a charm :)

**Comment by benpicco** [comment]
Hm, on saml21 I see the same power savings with #13475, that is just by using the IDLE mode.

**Comment by dylad** [comment]
@benpicco Please rebase.
For the record I tested this PR on saml11-xpro
examples/default on master
0.40 mA
examples/default with this PR
0.24 mA

**Comment by benpicco** [comment]
Also squash?

**Comment by dylad** [comment]
> Also squash?

Sure, feels free.

**Comment by benpicco** [comment]
Did you try `tests/periph_pm` and `set 0` on the saml11-xpro?
You will probably need #13475 for it to stay awake after leaving standby, but that's unrelated to this PR.

**Comment by dylad** [comment]
@benpicco:

```
2020-02-25 21:46:13,925 # main(): This is RIOT! (Version: 2020.04-devel-699-g7f4e3-cpu/sam0/use_buck_converter)
2020-02-25 21:46:13,931 # This application allows you to test the CPU power management.
2020-02-25 21:46:13,937 # The available power modes are 0 - 1. Lower-numbered power modes
2020-02-25 21:46:13,942 # save more power, but may require an event/interrupt to wake up
2020-02-25 21:46:13,945 # the CPU. Reset the CPU if needed.
2020-02-25 21:46:13,948 # using BTN0 as wake-up source
> set 0
2020-02-25 21:46:26,098 #  set 0
2020-02-25 21:46:26,101 # CPU will enter power mode 0.
```
power consumption at startup:
0.27 mA
after `set 0`
0.17 mA

**Comment by benpicco** [comment]
> 0.17 mA

hm that's underwhelming, saml21 does 1.8µA in standby - this one probably needs more work.

**Comment by dylad** [comment]
> hm that's underwhelming, saml21 does 1.8µA in standby - this one probably needs more work.

This PR still improves the current situation so this should belong to another PR.

**Comment by dylad** [comment]
@benpicco 
I just tested it on same54-xpro with your PR. My board doesn't want to boot at all.

**Comment by jue89** [comment]
> I just tested it on same54-xpro with your PR. My board doesn't want to boot at all.

The errata states that using the buck converter conflicts with the usage of PLLs (cf. [0] section 2.19.1) Maybe that's the reason ...

[0] http://ww1.microchip.com/downloads/en/DeviceDoc/SAM_D5x_E5x_Family_Errata_DS80000748J.pdf

**Comment by benpicco** [comment]
@jue89 thank you! You've saved me a lot of debugging. This is such a nasty bug, because if the `SEL` bit isn't cleared the buck converter will still be active after flashing a new firmware, so this morning I was thinking master was broken too!

Fortunately the `same54-xpro` does have a 12 MHz crystal oscillator that can be used as a clock source. With this we can enable the buck regulator again and get down to 0.8 mA in IDLE.

If the fast internal oscillators are used, the buck converter is only switched on during STANDBY and below where no fast oscillators are active. There it still provides ~5µA of current savings compared to LDO.

Now what I don't get yet is why it looks like I still get a 48 MHz clock from the 12 MHz crystal.

I was wondering whether it would fall back to the DFLL, but setting `USE_XOSC_ONLY` doesn't work on a custom board without the crystal, so it looks like it is used.
Well I guess we can still use the 26 MHz clock output from the `at86rf215` on that board, although that will require a new hardware revision. 

@dylad do you think this PR has gotten too messy? I can split it if you want.

File: boards/same54-xpro/include/periph_conf.h [hunk 3]
```diff
@@ -27,9 +27,26 @@ extern "C" {
...
- * @brief   GCLK reference speed
+ * @brief   To use the buck converter for best power efficiency, we must
+ *          use the external oscillator as a clock source.
+ *          DFLL and DPLL are not compatible with the buck regulator.
  */
+#define USE_XOSC_ONLY       (0)
+#define XOSC0_FREQUENCY     (12000000U)     /* 12 MHz */
+
+#if USE_XOSC_ONLY
+#define CLOCK_CORECLOCK     XOSC0_FREQUENCY
```
Comment (benpicco) [inline boards/same54-xpro/include/periph_conf.h]: Should probably be 48 MHz, but I don't yet know why.

**Comment by dylad** [comment]
> I can split it if you want.

Yes please, at least the newly XOSC part introduced for SAME54.

**Comment by benpicco** [comment]
Alright, I've move the samd5x external oscillator bits to it's own PR and only enabled the buck converter for SAML2x and SAML1x.

The buck converter is *always* incompatible with the fast internal clock generators, but since those are not used yet on these two families, we can still enable it by default.

File: cpu/samd5x/cpu.c [hunk 4]
```diff
@@ -64,7 +64,13 @@ static void dfll_init(void)
...
     ;
 
-    OSCCTRL->DFLLCTRLB.reg = reg;
+    /* workaround for Errata 2.8.3 DFLLVAL.FINE Value When DFLL48M Re-enabled */
+    OSCCTRL->DFLLMUL.reg = 0;   /* Write new DFLLMULL configuration */
+    OSCCTRL->DFLLCTRLB.reg = 0; /* Select Open loop configuration */
+    OSCCTRL->DFLLCTRLA.bit.ENABLE = 1; /* Enable DFLL */
+    OSCCTRL->DFLLVAL.reg = OSCCTRL->DFLLVAL.reg; /* Reload DFLLVAL register */
+    OSCCTRL->DFLLCTRLB.reg = reg; /* Write final DFLL configuration */
+
```
Comment (dylad) [inline cpu/samd5x/cpu.c]: I think this is unrelated to this PR and should be move in another PR.

Comment (benpicco) [reply cpu/samd5x/cpu.c]: Ok, dropped it.

File: cpu/sam0_common/include/periph_cpu_common.h [hunk 5]
```diff
@@ -378,13 +392,50 @@ static inline void sam0_cortexm_sleep(int deep)
...
 
+    cpu_pm_cb_leave(deep);
+
 #ifdef MODULE_PERIPH_GPIO
     gpio_pm_cb_leave(deep);
 #endif
 }
 
+/**
+ * @brief   Availiably supply controller voltage regulators.
```
Comment (dylad) [inline cpu/sam0_common/include/periph_cpu_common.h]: Availability  ?

File: cpu/sam0_common/include/periph_cpu_common.h [hunk 6]
```diff
@@ -378,13 +392,50 @@ static inline void sam0_cortexm_sleep(int deep)
...
+ */
+typedef enum {
+    SAM0_VREG_LDO,  /*< LDO, always availiable but not very power efficient */
```
Comment (dylad) [inline cpu/sam0_common/include/periph_cpu_common.h]: available ?

File: cpu/sam0_common/include/periph_cpu_common.h [hunk 7]
```diff
@@ -378,13 +392,50 @@ static inline void sam0_cortexm_sleep(int deep)
...
+    SAM0_VREG_BUCK  /*< Buck converter, efficient but may clash with internal
+                        fast clock generators (see errata sheets) */
+} sam0_supc_t;
+
+/**
+ * @brief       Switch the internal voltage regulator used for generating the
+ *              internal MCU voltages.
+ *              Availiable options are:
```
Comment (dylad) [inline cpu/sam0_common/include/periph_cpu_common.h]:
```suggestion
 *              Available options are:
```

File: cpu/samd5x/cpu.c [hunk 8]
```diff
@@ -144,11 +144,31 @@ uint32_t sam0_gclk_freq(uint8_t id)
...
+}
+
+void cpu_pm_cb_leave(int deep)
+{
+    if (deep) {
+        /* switch back to LDO */
+        sam0_set_voltage_regulator(SAM0_VREG_LDO);
+    }
+}
+
```
Comment (dylad) [inline cpu/samd5x/cpu.c]: Should we keep this in this PR ? I thought you only want to enable it for SAMLxx for now ?

File: cpu/saml21/cpu.c [hunk 9]
```diff
@@ -23,6 +23,14 @@
 #include "periph_conf.h"
 #include "stdio_base.h"
 
+/* As long as DFLL & DPLL are not used, we can default to
+   always using the buck converter when availiable. */
```
Comment (dylad) [inline cpu/saml21/cpu.c]:
```suggestion
   always using the buck converter when available. */
```

File: cpu/samd5x/cpu.c (hunk 8 above)
Comment (benpicco) [reply cpu/samd5x/cpu.c]: I noticed I can always enable the buck converter in standby and deep sleep, so I might as well do it together with enabling the buck converter for saml1x and saml2x.

I dropped the patch that fixes wake from standby when the DFLL is used, as this is unrelated as to which voltage regulator is used.
It requires the `cpu_pm_cb_leave()` callback though.

File: cpu/saml21/cpu.c [hunk 10]
```diff
@@ -82,6 +90,37 @@ uint32_t sam0_gclk_freq(uint8_t id)
...
+       to work around errata 1.2.1.
+       See discussion in #13441  */
+    assert((DSU->DID.bit.REVISION > 1) || ((PM->STDBYCFG.reg & 0x80) == 0));
+
+    /* errata 51.1.5 – When VDDCORE is supplied by the BUCK converter in performance
+                       level 0, the chip cannot wake-up from standby mode because the
+                       VCORERDY status is stuck at 0. */
+    if (USE_VREG_BUCK) {
+        sam0_set_voltage_regulator(SAM0_VREG_LDO);
+    }
```
Comment (dylad) [inline cpu/saml21/cpu.c]: SAML21 is set to PL2 in cpu_init, so I guess  it would be better to replace it by
```suggestion
    /* errata 51.1.5 – When VDDCORE is supplied by the BUCK converter in performance
                       level 0, the chip cannot wake-up from standby mode because the
                       VCORERDY status is stuck at 0. */
    if (USE_VREG_BUCK && !PM->PLCFG.bit.PLSEL) {
        sam0_set_voltage_regulator(SAM0_VREG_LDO);
    }
```

File: cpu/saml21/cpu.c [hunk 11]
```diff
@@ -23,6 +23,14 @@
 #include "periph_conf.h"
 #include "stdio_base.h"
 
+/* As long as DFLL & DPLL are not used, we can default to
+   always using the buck converter when available. */
+#if !defined(CPU_SAMR30)
+#define USE_VREG_BUCK   (1)
+#else
+#define USE_VREG_BUCK   (0)
+#endif
```
Comment (dylad) [inline cpu/saml21/cpu.c]: USE_VREG_BUCK should be a board parameter.
Some custom boards don't have the required coil to use VBUCK so unfortunately we cannot assign it at CPU level.
This is the same for every SAM0-based boards.

Comment (benpicco) [reply cpu/saml21/cpu.c]: Ah just like nrf52 - #13586
Do you know if there is any board supported by RIOT that *doesn't* have the coil? 
But looks like right now there are only `-xpro` boards with those MCUs so they are expected to have all the bells and whistles.

Comment (dylad) [reply cpu/saml21/cpu.c]: @benpicco 
Indeed, all Microchip devboards are equipped with the external coil. But we don't know all the boards in the wild so better take extra care there :)

Comment (benpicco) [reply cpu/saml21/cpu.c]: Yes I followed the example of `nrf5x` here.

**Comment by benpicco** [comment]
I would squash, rebase onto #10075 and add the new changes as fixups then, alright?

**Comment by dylad** [comment]
Sounds good to me !

**Comment by dylad** [comment]
Sounds good to me !

**Comment by benpicco** [comment]
There you go.

**Comment by benpicco** [comment]
replaced by #13764, #13786 & #13538
